src/crimson/os/seastore/async_cleaner.cc

```diff
@@ -706,38 +729,50 @@ SegmentCleaner::gc_cycle_ret SegmentCleaner::GCProcess::run()
     });
 }
 
-SegmentCleaner::gc_cycle_ret SegmentCleaner::do_gc_cycle()
+AsyncCleaner::gc_cycle_ret AsyncCleaner::do_gc_cycle()
 {
+  if (!gc_process.check_is_running()) {
+    gc_process.set_is_running();
```

Comment:
I think we need to root-cause why "do_gc_cycle is re-entered even though do_gc_cycle is already running" after this PR. That was impossible before this PR.

Reply:
In tm_single_device_test_t.random_writes_concurrent, transactions are issued concurrently as follows 

> DEBUG 2022-07-07 16:52:11,738 [shard 0] seastore_cache - 0x61600075d280 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 16:52:11), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=2, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 16:52:11,739 [shard 0] seastore_t - 0x61600075d580 Cache::create_transaction: created name=test_mutate, source=MUTATE, is_weak=false
DEBUG 2022-07-07 16:52:11,739 [shard 0] seastore_cache - 0x61600075d580 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 16:52:11), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=4, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 16:52:11,740 [shard 0] seastore_t - 0x61600075d880 Cache::create_transaction: created name=test_mutate, source=MUTATE, is_weak=false
DEBUG 2022-07-07 16:52:11,740 [shard 0] seastore_cache - 0x61600075d880 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 16:52:11), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=6, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 16:52:11,740 [shard 0] seastore_t - 0x61600075db80 Cache::create_transaction: created name=test_mutate, source=MUTATE, is_weak=false
DEBUG 2022-07-07 16:52:11,740 [shard 0] seastore_cache - 0x61600075db80 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 16:52:11), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=8, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 16:52:11,741 [shard 0] seastore_t - 0x61600075de80 Cache::create_transaction: created name=test_mutate, source=MUTATE, is_weak=false
DEBUG 2022-07-07 16:52:11,741 [shard 0] seastore_cache - 0x61600075de80 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 16:52:11), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=10, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 16:52:11,742 [shard 0] seastore_t - 0x61600075e180 Cache::create_transaction: created name=test_mutate, source=MUTATE, is_weak=false
DEBUG 2022-07-07 16:52:11,742 [shard 0] seastore_cache - 0x61600075e180 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 16:52:11), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=12, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 16:52:11,742 [shard 0] seastore_t - 0x61600075e480 Cache::create_transaction: created name=test_mutate, source=MUTATE, is_weak=false
DEBUG 2022-07-07 16:52:11,742 [shard 0] seastore_cache - 0x61600075e480 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 16:52:11), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=14, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 16:52:11,743 [shard 0] seastore_t - 0x61600075e780 Cache::create_transaction: created name=test_mutate, source=MUTATE, is_weak=false
DEBUG 2022-07-07 16:52:11,743 [shard 0] seastore_cache - 0x61600075e780 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 16:52:11), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=16, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 16:52:11,743 [shard 0] seastore_t - 0x61600075ea80 Cache::create_transaction: created name=test_mutate, source=MUTATE, is_weak=false
DEBUG 2022-07-07 16:52:11,743 [shard 0] seastore_cache - 0x61600075ea80 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 16:52:11), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=18, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 16:52:11,744 [shard 0] seastore_t - 0x61600075ed80 Cache::create_transaction: created name=test_mutate, source=MUTATE, is_weak=false
DEBUG 2022-07-07 16:52:11,744 [shard 0] seastore_cache - 0x61600075ed80 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 16:52:11), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=20, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 16:52:11,745 [shard 0] seastore_t - 0x61600075f080 Cache::create_transaction: created name=test_mutate, source=MUTATE, is_weak=false

Moreover, even though transactions has a conflict, the test eventually calls run_until_halt(). This results in there are in-flight gc tasks with conflicted extent. As a result, the test can hit `if (!extent.may_conflict())`, but Backref(Internal/Leaf)Node always return false when may_conflict is called unlike other extents.

So, Is this wrong situation? I thought that re-enter could occur.

Reply:
What I meant is that random_writes_concurrent() side, not the inside of do_gc_cycle.
1. run_until_halt() in try_submit_transaction() is called
2. context switched
3. another run_until_halt is called
4. ....

This run_until_halt calls do_gc_cycle.

Please refer to the following logs.

> DEBUG 2022-07-07 20:43:03,233 [shard 0] seastore_cleaner - AsyncCleaner::log_gc_state: caller GCProcess::run_until_halt, empty 117, open 5, closed 6, in_journal 0, total 1073741824B, available 1023381504B, unavailable 50360320B, unavailable_used 4304896B, unavailable_unused 46055424B; reclaim_ratio 0.9145180967873119, available_ratio 0.9530982971191406, should_block_on_trim false, should_block_on_reclaim false, gc_should_reclaim_space false, journal_head journal_seq_t(segment_seq=4489, offset=paddr_t<27570176>), journal_tail_target journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), journal_tail_commit journal_seq_t(segment_seq=0, offset=paddr_t<4096>), dirty_tail journal_seq_t(segment_seq=4485, offset=paddr_t<27570176>), dirty_tail_limit journal_seq_t(segment_seq=4481, offset=paddr_t<27570176>), gc_should_trim_journal true,
DEBUG 2022-07-07 20:43:03,233 [shard 0] seastore_t - 0x61600078d880 Cache::create_transaction: created name=trim_backref, source=TRIM_BACKREF, is_weak=false
DEBUG 2022-07-07 20:43:03,233 [shard 0] seastore_backref - 0x61600078d880 BtreeBackrefManager::merge_cached_backrefs: insert up to journal_seq_t(segment_seq=4485, offset=paddr_t<27570176>)
DEBUG 2022-07-07 20:43:03,233 [shard 0] seastore_backref - 0x61600078d880 BtreeBackrefManager::merge_cached_backrefs: seq journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), limit journal_seq_t(segment_seq=4485, offset=paddr_t<27570176>), num_fresh_backref 0
DEBUG 2022-07-07 20:43:03,233 [shard 0] seastore_backref - 0x61600078d880 BtreeBackrefManager::merge_cached_backrefs: new mapping: paddr_t<27537408>~4096 -> 0
DEBUG 2022-07-07 20:43:03,233 [shard 0] seastore_backref - 0x61600078d880 BtreeBackrefManager::new_mapping: 0~4096, paddr=paddr_t<27537408>
DEBUG 2022-07-07 20:43:03,233 [shard 0] seastore_cache - 0x61600078d880 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 20:43:02), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=512, user_hint=REWRITE, reclaim_gen=^A)
DEBUG 2022-07-07 20:43:03,233 [shard 0] seastore_cache - 0x61600078d880 Cache::get_root: after waiting root not on t --
DEBUG 2022-07-07 20:43:03,233 [shard 0] seastore_journal - 0x616000765380 ExtentPlacementManager::delayed_alloc_or_ool_write: start with 0 delayed extents

Context switch occurs here, then

> DEBUG 2022-07-07 20:43:03,235 [shard 0] seastore_cleaner - AsyncCleaner::log_gc_state: caller GCProcess::run_until_halt, empty 117, open 5, closed 6, in_journal 0, total 1073741824B, available 1023381504B, unavailable 50360320B, unavailable_used 4304896B, unavailable_unused 46055424B; reclaim_ratio 0.9145180967873119, available_ratio 0.9530982971191406, should_block_on_trim false, should_block_on_reclaim false, gc_should_reclaim_space false, journal_head journal_seq_t(segment_seq=4489, offset=paddr_t<27570176>), journal_tail_target journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), journal_tail_commit journal_seq_t(segment_seq=0, offset=paddr_t<4096>), dirty_tail journal_seq_t(segment_seq=4485, offset=paddr_t<27570176>), dirty_tail_limit journal_seq_t(segment_seq=4481, offset=paddr_t<27570176>), gc_should_trim_journal true,
DEBUG 2022-07-07 20:43:03,235 [shard 0] seastore_t - 0x61600078db80 Cache::create_transaction: created name=trim_backref, source=TRIM_BACKREF, is_weak=false
DEBUG 2022-07-07 20:43:03,235 [shard 0] seastore_backref - 0x61600078db80 BtreeBackrefManager::merge_cached_backrefs: insert up to journal_seq_t(segment_seq=4485, offset=paddr_t<27570176>)
DEBUG 2022-07-07 20:43:03,235 [shard 0] seastore_backref - 0x61600078db80 BtreeBackrefManager::merge_cached_backrefs: seq journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), limit journal_seq_t(segment_seq=4485, offset=paddr_t<27570176>), num_fresh_backref 0
DEBUG 2022-07-07 20:43:03,235 [shard 0] seastore_backref - 0x61600078db80 BtreeBackrefManager::merge_cached_backrefs: new mapping: paddr_t<27537408>~4096 -> 0
DEBUG 2022-07-07 20:43:03,235 [shard 0] seastore_backref - 0x61600078db80 BtreeBackrefManager::new_mapping: 0~4096, paddr=paddr_t<27537408>
DEBUG 2022-07-07 20:43:03,235 [shard 0] seastore_cache - 0x61600078db80 Cache::get_root: root not on t -- CachedExtent(addr=0x61a000554480, type=ROOT, version=820, dirty_from_or_retired_at=journal_seq_t(segment_seq=4484, offset=paddr_t<27533312>), modify_time=tp(2022-07-07 20:43:02), paddr=paddr_t<NULL>, length=0, state=DIRTY, last_committed_crc=0, refcount=514, user_hint=REWRITE, reclaim_gen=^A)

There is another run_until_halt()

Reply:
Ahh, good catch! But I think there is no need to complicate the non-test path as it is only possible through concurrent calls from UT to `GCProcess::run_until_halt()`. Probably just prevent that in the `GCProcess::run_until_halt()` rather than in the common `do_gc_cycle()` and `gc_should_run()`?

Reply:
OK.